Add InteractiveTools.

blankenberg opened the pull request:
Basic idea: launch a container-backed Galaxy Tool and get access to content inside in real time.

- Give user access via uwsgi (inspired by https://github.com/galaxyproject/galaxy/pull/2385), or another proxy.
- Access is based upon a key, key_type, token mapping to a host, port. You can have and access any number of RealTimeTools at a time. 
- Port and entry url-path snippet are specified in standard tool.xml files. 
- You can specify any number of ports, so a single RealTimeTool can give access to multiple running applications.
- RealTimeTools can be added to and installed from the ToolShed.
- Currently working for docker in local runner.

blankenberg committed:
Add RealTimeTools. Currently working for docker in local runner.

blankenberg committed:
Flake8

nsoranzo commented on lib/galaxy/datatypes/realtime.py:
```diff
@@ -0,0 +1,77 @@
+import os
+import sys
+
+from text import Html
```
```suggestion
from .text import Html
```

nsoranzo committed:
Update lib/galaxy/datatypes/realtime.py

Co-Authored-By: blankenberg <dan.blankenberg@gmail.com>

blankenberg committed:
fix

blankenberg committed:
fix

blankenberg committed:
Make stop button in Active RealTimeTool list do something.

blankenberg committed:
better handling of missing containers

blankenberg committed:
merge dev into realtimetools

blankenberg committed:
Update access to url_for().

blankenberg committed:
import ordering

blankenberg committed:
Only display view result message when view is available.

blankenberg committed:
Add a ttl cache aware mapper to key_type_token_mapping.py.

blankenberg committed:
Remove entries from realtimetool sqlite db when job has ended.

blankenberg committed:
Simplify RealTimeTool model, associated entry points directly with job.

blankenberg committed:
merge dev

blankenberg committed:
remove debug statement

blankenberg committed:
merge dev

blankenberg committed:
comma

blankenberg committed:
Merge branch 'dev' of https://github.com/galaxyproject/galaxy into realtimetools

blankenberg committed:
fix for displaying realtimetool grid on refresh.

blankenberg committed:
Merge branch 'dev' of https://github.com/galaxyproject/galaxy into realtimetools

blankenberg committed:
Ports for yaml tools.

blankenberg committed:
update tools xsd

blankenberg committed:
Merge branch 'dev' of https://github.com/galaxyproject/galaxy into realtimetools

blankenberg committed:
over zealous delete key

blankenberg committed:
flake

blankenberg committed:
flake

blankenberg committed:
naming update

blankenberg committed:
move realtime rpc script to /scripts/realtime/

blankenberg committed:
move realtime rpc script to /scripts/realtime/

blankenberg committed:
Revert "move realtime rpc script to /scripts/realtime/"

This reverts commit 9ee4b9c4e293099be5728ef0ac155309fe2bd1e7.

blankenberg committed:
rename rtt db migration

blankenberg committed:
Merge branch 'dev' of https://github.com/galaxyproject/galaxy into realtimetools

blankenberg committed:
Merge branch 'dev' of https://github.com/galaxyproject/galaxy into realtimetools

blankenberg committed:
rename realtimetool migration to higher version

blankenberg committed:
Merge branch 'dev' of https://github.com/galaxyproject/galaxy into realtimetools

blankenberg committed:
Merge branch 'dev' of https://github.com/galaxyproject/galaxy into realtimetools

blankenberg committed:
change version number of realtimetools migration

blankenberg committed:
Merge branch 'dev' of https://github.com/galaxyproject/galaxy into realtimetools

blankenberg committed:
Merge branch 'dev' of https://github.com/galaxyproject/galaxy into realtimetools

blankenberg committed:
update migrate version of realtimetools

blankenberg committed:
Merge branch 'dev' of https://github.com/galaxyproject/galaxy into realtimetools

jmchilton commented:
Committers meeting sort of got hung up on this, we need more context here I think @blankenberg. Can you provide an example in tree and outline the compelling use case driving this development?

jmchilton commented:
I'm going to push this to 19.09 because @jxtx marked himself as needing to review it, I'm likewise not ready to merge it. We have a week, it may still get merged.

blankenberg committed:
Merge branch 'dev' of https://github.com/galaxyproject/galaxy into realtimetools

blankenberg committed:
Remove inspection check, only do ports

blankenberg committed:
Remove unneeded RealTimeTool datatype.

blankenberg committed:
No more inspect, no more json needed.

jmchilton commented:
I'm 👍 on the idea but I'm not quite sold on the implementation yet, however I will be happy to do the leg work to address my own concerns. My intention is:

- Integrate this Pulsar - develop an internal API to publish hostname/port back to Galaxy.
- Implement a notebook tool.
- Add a Selenium test case.

This is to convince myself this generalizes beyond the local job runner, we can support it in an ongoing fashion, we have the right data model, and to see if I can avoid persisting those docker commands in the job state.

I'd also like to see the proxies merged, but that seems unlikely at first glance. I'd also like to see it hidden behind ``enable_beta_tool_formats`` for one release, but I won't insist on it.

If I have not made substantial progress on these goals in the next month (ping me May 16th), I will merge it as is barring comments from other reviewers.

natefoo commented:
I'd also like to see if we can get uWSGI to manipulate the paths of proxied cookies, this will help make them work out of the box without needing a DNS wildcard and a wildcard SSL cert.

bgruening commented on lib/galaxy/jobs/runners/local.py:
```diff
@@ -170,6 +181,7 @@ def stop_job(self, job_wrapper):
                 return
         else:
             log.warning("stop_job(): %s: PID %d refuses to die after signaling TERM/KILL" % (job.id, pid))
+        self.__kill_container(job_wrapper)
```
`self._kill_container(job_wrapper)`

blankenberg replied:
Nice catch, thanks!

bgruening commented on lib/galaxy/jobs/runners/local.py:
```diff
@@ -213,12 +225,70 @@ def _check_pid(self, pid):
+                        ports_raw = None
+                        for i in range(1, max_command_attempts):
+                            with tempfile.TemporaryFile() as stdout_file:
+                                exit_code = subprocess.call(cont.container_info['commands']['port'],
```
Why not `_run_command`?

blankenberg replied:
We need to capture and parse `stdout` to get the port info here, whereas in the other cases, we just let the stream pass through. Don't want the overhead of making a temp file when we don't need it. Could make it flag-able in the method, but it might be messier due to the with.

bgruening committed:
change cellxgene Container to work with the IE tools

bgruening committed:
make entry_point_url parsing more robust

bgruening committed:
add initial condor runner support

bgruening committed:
add initial test tools for a interactive Vis

bgruening committed:
add initial Jupyter notebook tool, this tool supports loading a notebook and execute it automatically, with that it can run in a workflow

bgruening committed:
fix bug in setting the --net option during Docker startup

blankenberg committed:
Merge branch 'dev' of https://github.com/blankenberg/galaxy into realtimetools

blankenberg committed:
Update rtt migrate version number

blankenberg committed:
js-lint fix

bgruening reviewed (commented):
Hi Dan, I added a bunch of comments and

* documentation
* condor-runner integration, yeah it works! :)
* 2 IEs that demonstrate the usage, Jupyter and Single-Cell

https://github.com/blankenberg/galaxy/pull/4/files

Everything rough, sorry we do not have that much time at the moment :(

blankenberg committed:
remove extra _

blankenberg committed:
Merge branch 'dev' of https://github.com/galaxyproject/galaxy into realtimetools

blankenberg commented:
> Hi Dan, I added a bunch of comments and
> 
> * documentation
> * condor-runner integration, yeah it works! :)
> * 2 IEs that demonstrate the usage, Jupyter and Single-Cell
> 
> https://github.com/blankenberg/galaxy/pull/4/files
> 
> Everything rough, sorry we do not have that much time at the moment :(

Awesome news. Will take a look soon. Thanks!

blankenberg committed:
Merge branch 'dev' of https://github.com/galaxyproject/galaxy into realtimetools

blankenberg committed:
Update RealTimeTool entry point definitions.

blankenberg committed:
Update xsd

blankenberg committed:
Enable templating of RealTime tools name, port, url.

blankenberg committed:
Add docs and example job config.

blankenberg committed:
Merge branch 'dev' of https://github.com/galaxyproject/galaxy into realtimetools

blankenberg committed:
Remove unused import

blankenberg committed:
Update example RealTimeTools to use current syntax.

blankenberg committed:
Merge pull request #5 from blankenberg/realtimetools-bg

Realtimetools bg

blankenberg committed:
flake8 for lib/galaxy/jobs/runners/condor.py

blankenberg committed:
mock

blankenberg committed:
Update docs

blankenberg committed:
Ports to test.

blankenberg committed:
Doc formatting

blankenberg committed:
More doc info

blankenberg committed:
notes on resource consumption; and condor

blankenberg committed:
Merge branch 'dev' of https://github.com/galaxyproject/galaxy into realtimetools

blankenberg committed:
update docs

hexylena commented:
Hi @blankenberg, you suggested we continue the discussion here

> > I think we should stick to the "Interactive" name and keep that branding in the community and migrate all internal IEs that are feasible to those tools, this could be a nice hackathon project 
>
> I disagree, these are Galaxy RealTimeTools. GIEs are something else, as described in 10.1371/journal.pcbi.1005425.

Likewise, I disagree. You implement the core ideas of GIEs and things that were on our roadmap, while positioning this as something novel. It's features we've discussed previously that we either wanted to implement or wanted to discuss with the devteam what the best implementation would be.

- [where it was raised that we should run these as tools](https://github.com/galaxyproject/galaxy/issues/1923) which we merged into our GIE roadmap but had temporarily removed from our high priority GIE fixes
- [managing multiple running containers](https://github.com/galaxyproject/galaxy/issues/4651) and long running containers
- [ability to shut down IEs when done](https://github.com/galaxyproject/galaxy/issues/1586)

In the most optimistic interpretation of your actions, how this was done was a bit tone deaf and without considerations for optics or community contributors.

I'm struggling not to have a more negative interpretation, perhaps you can clear this up for me?

blankenberg committed:
More doc formatting

blankenberg committed:
Merge branch 'dev' of https://github.com/galaxyproject/galaxy into realtimetools

blankenberg committed:
missing t

nekrut commented:
It pains me greatly to see this level of deliberations related to naming. We can probably dig into the exact history from the start here using github, but I'm pretty sure the first GIEs were developed by Bjorn and Helena. They probably weren't the first one who though of doing interactive analyses in Galaxy, but they did implement this (I certainly though of creating infinite energy using cold fusion, but it is unlikely that my name will be associated with it, which is perfectly fine with me). This got all of us excited and this is how the GIE manuscript came about. The future of Galaxy is very much dependent on ability to seamlessly switch between Galaxy proper and *ad hoc* analysis environments such as Jupyter, RStudio, and anything else, as well as interactive visualizations. This was a great breakthrough, but as time went by we realized that just as piston engines were not good for supersonic flight the original GIEs could be improved on and better integrated with Galaxy. This is true for any first implementation. Dan rethought this and developed iteration 2 of this. It may be mechanistically different, but ultimately it achieves the same goal. So a good compromise will be to just call it something like "interactive tools" and get this settled. This does not in any possible way diminish one's ability to publish on that. In fact, the paper will be significantly better if it embraces all community members who care about GIEs.

bgruening commented:
Hi all,

I would like to keep this PR technical, lets move the historical and political aspect to a chat with all interesting parties and figure out what is our next plans.
I think there is and was a huge amount of miscommunication involved here and I would like to propose an IE/RRT meeting. Would you all have time on Tuesday at 16.00 CEST? We can use gitter as some indicated that text would be better than a video call.

Would be great if you all can make it.

jmchilton commented:
The two things that I grate on most here when digging through the code are writing command-lines to the database and the nested loops and conditionals in the runner code that are tied to particular container types and these commands to kill the container.

I had a thought about a different, potentially more general approach. I feel like the command builder is already dealing with container details, what if we wrote the stop/kill commands into the job script with a trap https://www.shellscript.sh/trap.html function. Would that solve the problem with the stop/kill command for all the runners and for all container execution? Does anyone know off hand if that is going to work (or not) before I try it out?

I'm not suggesting that this should necessarily block the merge of this - because I'm willing to accept containerized tools are things I committed and were more or less broken by design, but if we're going to fix it I wonder if we should fix it more generally and fixing it would make this more general and make more more comfortable with what is here.

jmchilton commented on test/functional/tools/realtimetool_cellxgene.xml:
```diff
@@ -0,0 +1,27 @@
+        </entry_point>
+    </realtime>
+    <command><![CDATA[
+        cp '${input}' '/tmp/galaxy_cellxgene_${input.element_identifier}.h5ad'
```
I'm sure you hate the best practice for dealing element_identifier Dan 😓, but please clean it the way @bgruening's tool does.

xref https://planemo.readthedocs.io/en/latest/writing_advanced.html#processing-identifiers
xref #7937

blankenberg replied:
I mean sure, I will fix this, but this was from the code that @bgruening PR'd, just cleaned up a little bit with adding quotes after the merge conflict.

blankenberg replied:
re.sub'd in 87fed7e8eede87a6ab97e1f0862a9727d9a19880

bgruening replied:
Please treat all what I did as experimental. I just wanted to demonstrate that this kind of works for Jupyter, RStudio and where are limitations.

jmchilton committed:
Simplify set_name stuff in container.

jmchilton commented:
Cherry-pick https://github.com/galaxyproject/galaxy/commit/3d2b6bdec54fb94c1480777b39e3f4ab55b5220e?

I really like JobContainerAssociation and I like the idea of just assigning all containers a name.

blankenberg committed:
Merge branch 'dev' of https://github.com/galaxyproject/galaxy into realtimetools

jmchilton commented on lib/galaxy/tools/deps/container_classes.py:
```diff
@@ -263,12 +273,17 @@ def containerize_command(self, command):
             env_directives=env_directives,
             working_directory=working_directory,
-            net=self.prop("net", "none"),  # By default, docker instance has networking disabled
+            net=self.prop("net", None),  # By default, docker instance has networking disabled
```
This change seems wrong? Won't this disable disabling network access by default or was there a bug here already?

blankenberg replied:
@bgruening was fixing a bug?

bgruening replied:
I think this fixed a bug in my testing. The former yielded to `--net=None` I think and not to ` ` which was expected.

blankenberg committed:
re.sub on cellxgene input name

blankenberg commented:
> Cherry-pick [3d2b6bd](https://github.com/galaxyproject/galaxy/commit/3d2b6bdec54fb94c1480777b39e3f4ab55b5220e)?

Done and done, also you have far too many branches...

jmchilton commented:
I really like tracking the container job association and thanks for cherry picking what I sent you Dan. That said, I think I dislike writing ``container_info``, modifying it, re-writing it. It seems to violate abstractions at that high level description, and digging into it - ``containerize_command`` modifying ``container_info`` seems odd and a violation also. That method should not be mutating state in my opinion and that side effect is non-obvious. It seems like as long as we make sure ``docker_host_props`` are preserved (maybe call it ``container_target_props`` for compatibility with what that would look like in Singularity)  on that container association we can get away without doing all of that:

So instead of pre-computing commands we may need for every job and violating the abstractions:

```python
        self.container_info['commands'] = {}
        for cmd in ['inspect', 'port', 'kill', 'stop']:
            self.container_info['commands'][cmd] = docker_util.build_docker_simple_command(cmd, container_name=self.container_name, **docker_host_props)
```

We could just add something like this to job_wrapper:

```python
     def container_command_ports(self):
          self._container_command('ports')

     def _container_command(self, command):
           container_assoc = self.get_job().container  # container_assoc
           return docker_util.build_docker_simple_command(command, container_assoc.container_name, container_assoc.container_target_props)
````

I'm nervous this following line was really confusing and involved changing a property with complex state already:

```python
        stored_container = job_wrapper.get_job().container
        if container and stored_container:
            if container.container_info:
                stored_container.container_info = container.container_info
```

It would become the following, which is clearer to me and which avoids re-writing container_info. We're writing a different property that didn't previously have state and only gets written in one place.

```python
        # Make sure target container is written for command generation.
        stored_container = job_wrapper.get_job().container
        if container and stored_container:
              stored_container.container_target_props = container.container_target_props
```

Before actually trying this out I was wondering how do these abstraction tweaks sound to you?

blankenberg commented:
> Before actually trying this out I was wondering how do these abstraction tweaks sound to you?

Abstractions are good :) Certainly makes sense to build those container interaction commands on the fly.

blankenberg committed:
Only set container if container exists, fixes tests failing due to Mock object.

jmchilton committed:
Integration testing for RTT API hooks.

jmchilton commented:
https://github.com/galaxyproject/galaxy/commit/2876d48a88d1d0552bacbf64f39eabbaa9819218 Integration tests for this... too much depends on web controllers instead of API controllers to make more progress without either...

- Adding a new test suite that blends per-testcase Galaxy configs of integration testing with configuration of Selenium environment (a project I've wanted to do for a long time - found this WIP branch from a year ago - https://github.com/jmchilton/galaxy/tree/selenium_integration - one of those way too many branches). This test suite would really let us test whole new classes of awesome things we just cannot test currently and is very important.
- ***or*** Refactoring more this PR to work with the API controllers. Again this is probably a good idea anyway - we've wanted to avoid adding new non-API controllers for years now and generally push back against them.

I'll try to make some progress on one or the other.

blankenberg committed:
Merge branch 'dev' of https://github.com/galaxyproject/galaxy into realtimetools

jmchilton committed:
Refactor realtime entry point handling toward API usage.

- Add API entry point for fetching a target from an encoded entry point ID.
- Refactor target generation into a manager.
- Use galaxy.exceptions and exception handling instead of trans.show_error_message directly.
- Extend grid view to leverage this new API and use single-page error handling.

jmchilton commented:
Following WIP commit fetches that entry point URI target from the API instead of generating it in the controller. I didn't quite understand what the index controller was doing and overshot the amount of controller code that can be deleted... but it is an approach I think we should work toward.

https://github.com/jmchilton/galaxy/commit/8b048095734b7170ddc9f3deb4268c6bac6c8f8d

Error handling will be more SPA-y and more functionality will be accessible to developers via the API - hopefully these are both appealing.

Tomorrow I'll see if I can get it working and then get that integration test working based on it, and then try to apply some refactorings I discussed above and some I have stuffed in local branches.

blankenberg committed:
Merge branch 'dev' of https://github.com/galaxyproject/galaxy into rtt-tmp

blankenberg committed:
Fix tool_util import

stupid .pyc files

jmchilton committed:
Move computing result_view for RTT to client.

jmchilton commented:
VueJs component for polling status of RTT right on the execution result page, progress away from backbone and web controllers and toward Vue and API controllers...

https://github.com/jmchilton/galaxy/commit/7bcb1dfcc82f632b2c133a032465091cd1b9d326

blankenberg commented:
Awesome @jmchilton. Especially like the updating message on the tool submit success form (clicking did not work for me, however).

One thing that would be nice would be if the links were 'real' and allowed optionally opening into new tabs, etc with ctrl- or right-click.

jmchilton commented:
So... this only works in Chrome for me? I can generate one of your ... interesting... realtime proxy URLs and then use it in that tab or in an incognito browser for Chrome. But I can't use these URLs with wget, Python requests, Firefox, or Safari. @natefoo was skeptical about this localhost wildcard prefix trick - but maybe it only works in Chrome? I guess with a willcard actual DNS record this would work regardless of the Chrome magic?

Updated:

Work around for Python testing...

```
>>> requests.get("http://localhost:8080", headers={"Host": "realtimetoolentrypoint.9ce617df390851fc.b529bca87f604f559994b1e0b0848435.realtime.localhost:8080"}).content
'moo cow\n'```

natefoo commented:
Wildcard DNS is going to break most people's SSL setups though. See my comment at the top about cookie path manipulation.

blankenberg commented:
> So... this only works in Chrome for me? I can generate one of your ... interesting... realtime proxy URLs and then use it in that tab or in an incognito browser for Chrome. But I can't use these URLs with wget, Python requests, Firefox, or Safari. @natefoo was skeptical about this localhost wildcard prefix trick - but maybe it only works in Chrome? I guess with a willcard actual DNS record this would work regardless of the Chrome magic?
> 
> Updated:
> 
> Work around for Python testing...
> 
> ```
> >>> requests.get("http://localhost:8080", headers={"Host": "realtimetoolentrypoint.9ce617df390851fc.b529bca87f604f559994b1e0b0848435.realtime.localhost:8080"}).content
> 'moo cow\n'```
> ```

Well, (I think) it should work in all:
https://tools.ietf.org/html/rfc2606
```
      The ".localhost" TLD has traditionally been statically defined in
      host DNS implementations as having an A record pointing to the
      loop back IP address and is reserved for such use.  Any other use
      would conflict with widely deployed code which assumes this use.
```
And there is: https://tools.ietf.org/html/rfc6761
And https://tools.ietf.org/html/draft-west-let-localhost-be-localhost-06

But in just testing now, it worked for me in Opera and Chrome, but not Firefox or Safari.

For subdomains with wildcard A records, it was working for each browser that I tested. I did not test SSL certs, but it is possible to get wildcard SSL certs.

jmchilton committed:
Change default realtime prefix to something that isn't broken.

blankenberg commented:
FWIIW: This is the change where I first swapped from using a path-based approach to subdomain: https://github.com/galaxyproject/galaxy/pull/7494/commits/a7c9ac042f41c9b0b416875b8cc1c676b5af30b9

It solved a bunch of issues I was having when testing features of various RealTimeTool example cases, and also seemed like a potentially good way to help with distributed-routing in the future.

With just a little more effort the path-based approach could likely be made to work well.

jmchilton committed:
Rebase into intergration testsing..

jmchilton commented:
> One thing that would be nice would be if the links were 'real' and allowed optionally opening into new tabs, etc with ctrl- or right-click.

Ugh... yeah that makes sense but I can't do that cleanly from a click handler. I guess I'll make the grid active and poll the entry points and only enable the link when it becomes active.

jmchilton committed:
Rework client polling & entry point index API for potential reuse in the grid.

Fix links in the job monitor widget.

jmchilton commented:
1126d68fedc32c1644915addce951a945496d776 reworking toward that polling, pre-computing those links and placing them in the main API response.

I don't know about these links to be honest, they make me quite nervous. This proxy doesn't require the Galaxy cookie so those links feel less secure than GIE approach - I'm not sure if that is just a feeling though of if there is something to that 🤔. To get the user's API key from a GIE notebook, you need both the link to the session and the cookie I think... not just some permanent link that might be saved in a browser history or obtained from an XSS attack. Less relevant for tools that just produce single outputs though obviously...

jmchilton committed:
Make realtime tool grid actual links, poll on active.

jmchilton committed:
Realtime Tools: Monitor for entrypoint ports in job script.

A few more pieces for local execution but more general, only requires a shared file system with Galaxy instead of direct access to docker host the job is running inside. Allows us to skip populating 'port' command in job destination. Eliminates a lot of condor hackiness - that implementation shares a lot more code with the local runner now.

jmchilton committed:
Proper container killing when running jobs.

Rip out container_info container hacks for interactive tools, container killing should now just work with local job runner (or any job runners).

jmchilton commented:
- Proper container killing that should work across all job runners, removing the local runner code and container_info hacking I disliked the most  https://github.com/galaxyproject/galaxy/pull/7494/commits/70989ef32694dfbbd4bc3e63b0ac72624f004fb8.
- Delegate detecting ports to the job script, should make it much more easy to generalize this to other runners https://github.com/galaxyproject/galaxy/pull/7494/commits/0dfe1343681db04223c4f5696649a2a5dca0c9a4

jmchilton committed:
Add missing container_monitor package to previous commits.

blankenberg committed:
Merge branch 'dev' of https://github.com/galaxyproject/galaxy into realtimetools

blankenberg committed:
Lowercase import Vue from "vue";

blankenberg committed:
flake8

blankenberg committed:
jslint

blankenberg committed:
flake

blankenberg committed:
Update driver_util.py

blankenberg committed:
Add container_montior_command to MockJobWrapper

blankenberg committed:
montior --> monitor

blankenberg committed:
juypter --> jupyter

blankenberg committed:
Merge branch 'dev' of https://github.com/galaxyproject/galaxy into realtimetools

blankenberg committed:
Merge branch 'dev' of https://github.com/galaxyproject/galaxy into realtimetools

blankenberg committed:
Merge branch 'dev' of https://github.com/galaxyproject/galaxy into realtimetools

blankenberg committed:
Change RealTimeTool to InteractiveTool in UI.

blankenberg committed:
Change realtime tagset to entry_points.

blankenberg committed:
Merge pull request #7 from blankenberg/interactivetools

Rename RealTimeTools to InteractiveTools

blankenberg committed:
interactive

blankenberg committed:
Fix for check_for_entry_points when file exists, but is not fully populated yet

blankenberg committed:
Pass through tool defined environment_variables to Docker.

blankenberg committed:
Merge branch 'dev' of https://github.com/galaxyproject/galaxy into realtimetools

blankenberg committed:
from galaxy.web.base.controller import is now from galaxy.webapps.base.controller import

blankenberg committed:
Add 404 when InteractiveTool route doesn't exist or can't be accessed

blankenberg committed:
Flake

blankenberg committed:
Add ability to strip whitespace from tool environment_variables.

blankenberg committed:
Fix for fetching strip if not defined

blankenberg committed:
Merge branch 'dev' of https://github.com/galaxyproject/galaxy into realtimetools

rdallet commented:
Hi Dan,

I have developed several Galaxy Interactive Environments and I am therefore very interested in Interactive Tools. I tested the latest version of your Pull Request but I get this error.

```
cp: cannot stat ‘/home/rdallet/galaxy/lib/galaxy/tools/test/functional/tools/default_notebook.ipynb’: No such file or directory
```

I have this file in `/home/rdallet/galaxy/test/functional/tools/default_notebook.ipynb` but not in `lib/galaxy/tools/...`. How can I fix this ?

blankenberg commented:
> Hi Dan,
> 
> I have developed several Galaxy Interactive Environments and I am therefore very interested in Interactive Tools. I tested the latest version of your Pull Request but I get this error.
> 
> ```
> cp: cannot stat ‘/home/rdallet/galaxy/lib/galaxy/tools/test/functional/tools/default_notebook.ipynb’: No such file or directory
> ```
> 
> I have this file in `/home/rdallet/galaxy/test/functional/tools/default_notebook.ipynb` but not in `lib/galaxy/tools/...`. How can I fix this ?

Hi Romain,

Is this error when you are trying to use the example Jupyter notebook InteractiveTool found at `test/functional/tools/realtimetool_juypter_notebook.xml`, or your own notebook?

Thanks,

Dan

rdallet commented:
> > Hi Dan,
> > I have developed several Galaxy Interactive Environments and I am therefore very interested in Interactive Tools. I tested the latest version of your Pull Request but I get this error.
> > ```
> > cp: cannot stat ‘/home/rdallet/galaxy/lib/galaxy/tools/test/functional/tools/default_notebook.ipynb’: No such file or directory
> > ```
> > 
> > 
> > I have this file in `/home/rdallet/galaxy/test/functional/tools/default_notebook.ipynb` but not in `lib/galaxy/tools/...`. How can I fix this ?
> 
> Hi Romain,
> 
> Is this error when you are trying to use the example Jupyter notebook InteractiveTool found at `test/functional/tools/realtimetool_juypter_notebook.xml`, or your own notebook?
> 
> Thanks,
> 
> Dan

The Jupyter notebook InteractiveTool example. (`test/functional/tools/realtimetool_juypter_notebook.xml`)

blankenberg commented:
> The Jupyter notebook InteractiveTool example. (`test/functional/tools/realtimetool_juypter_notebook.xml`)

And you are loading the tool directly, or have you copied it?

From the xml, it calls
`cp '$__tool_directory__/default_notebook.ipynb' ./ipython_galaxy_notebook.ipynb &&`

where `$__tool_directory__` refers to the directory where the xml file is located, which should resolve to `test/functional/tools/default_notebook.ipynb`.

rdallet commented:
> And you are loading the tool directly, or have you copied it?

I cloned your PR and pulled the latest version. I didn't change anything exept the config files (config/galaxy.yml , config/job_conf.xml, config/tool_conf.xml).

> From the xml, it calls
> `cp '$__tool_directory__/default_notebook.ipynb' ./ipython_galaxy_notebook.ipynb &&`
> 
> where `$__tool_directory__` refers to the directory where the xml file is located, which should resolve to `test/functional/tools/default_notebook.ipynb`.

OK, so `default_notebook.ipynb` is in `/home/rdallet/galaxy/test/functional/tools/`, but it's like `$__tool_directory__` refers to `/home/rdallet/galaxy/lib/galaxy/tools/test/functional/tools/` instead of `/home/rdallet/galaxy/test/functional/tools/`. Can be a configuration problem?

blankenberg commented:
> OK, so `default_notebook.ipynb` is in `/home/rdallet/galaxy/test/functional/tools/`, but it's like `$__tool_directory__` refers to `/home/rdallet/galaxy/lib/galaxy/tools/test/functional/tools/` instead of `/home/rdallet/galaxy/test/functional/tools/`. Can be a configuration problem?

Maybe a general configuration problem? How are you loading the test tool? I am somewhat lazy, so in the `tool_conf.xml` file in my dev instance I just have:
```
  <section id="functest" name="Functional Test Tools">
    <tool_dir dir="/path/to/git/galaxy/test/functional/tools" />
  </section>
```
which loads all of the test tools and the jupyter InteractiveTool starts up OK for me.

rdallet commented:
OK so it was a configuration problem. I used the documentation in `doc/source/admin/special_topics/` and I added `<tool file="../test/functional/tools/realtimetool_juypter_notebook.xml" />` in `tool_conf.xml`.
The problem was solved when I changed the tool file path to an absolute path. But I don't understand why the relative path doesn't work.

And now I have a second issue... This seems to be a docker one.
```
/opt/conda/lib/python3.5/site-packages/IPython/paths.py:61: UserWarning: IPython dir '/home/jovyan/.ipython' is not a writable location, using a temp directory.
  " using a temp directory.".format(ipdir))
Traceback (most recent call last):
  File "/opt/conda/bin/jupyter-trust", line 6, in <module>
    sys.exit(nbformat.sign.TrustNotebookApp.launch_instance())
  File "/opt/conda/lib/python3.5/site-packages/jupyter_core/application.py", line 267, in launch_instance
    return super(JupyterApp, cls).launch_instance(argv=argv, **kwargs)
  File "/opt/conda/lib/python3.5/site-packages/traitlets/config/application.py", line 657, in launch_instance
    app.initialize(argv)
  File "<decorator-gen-6>", line 2, in initialize
  File "/opt/conda/lib/python3.5/site-packages/traitlets/config/application.py", line 87, in catch_config_error
    return method(app, *args, **kwargs)
  File "/opt/conda/lib/python3.5/site-packages/jupyter_core/application.py", line 243, in initialize
    self.migrate_config()
  File "/opt/conda/lib/python3.5/site-packages/jupyter_core/application.py", line 169, in migrate_config
    migrate()
  File "/opt/conda/lib/python3.5/site-packages/jupyter_core/migrate.py", line 241, in migrate
    with open(os.path.join(env['jupyter_config'], 'migrated'), 'w') as f:
PermissionError: [Errno 13] Permission denied: '/home/jovyan/.jupyter/migrated'
/opt/conda/lib/python3.5/site-packages/IPython/paths.py:61: UserWarning: IPython dir '/home/jovyan/.ipython' is not a writable location, using a temp directory.
  " using a temp directory.".format(ipdir))
Traceback (most recent call last):
  File "/opt/conda/bin/jupyter-notebook", line 6, in <module>
    sys.exit(notebook.notebookapp.main())
  File "/opt/conda/lib/python3.5/site-packages/jupyter_core/application.py", line 267, in launch_instance
    return super(JupyterApp, cls).launch_instance(argv=argv, **kwargs)
  File "/opt/conda/lib/python3.5/site-packages/traitlets/config/application.py", line 657, in launch_instance
    app.initialize(argv)
  File "<decorator-gen-7>", line 2, in initialize
  File "/opt/conda/lib/python3.5/site-packages/traitlets/config/application.py", line 87, in catch_config_error
    return method(app, *args, **kwargs)
  File "/opt/conda/lib/python3.5/site-packages/notebook/notebookapp.py", line 1290, in initialize
    super(NotebookApp, self).initialize(argv)
  File "<decorator-gen-6>", line 2, in initialize
  File "/opt/conda/lib/python3.5/site-packages/traitlets/config/application.py", line 87, in catch_config_error
    return method(app, *args, **kwargs)
  File "/opt/conda/lib/python3.5/site-packages/jupyter_core/application.py", line 243, in initialize
    self.migrate_config()
  File "/opt/conda/lib/python3.5/site-packages/jupyter_core/application.py", line 169, in migrate_config
    migrate()
  File "/opt/conda/lib/python3.5/site-packages/jupyter_core/migrate.py", line 241, in migrate
    with open(os.path.join(env['jupyter_config'], 'migrated'), 'w') as f:
PermissionError: [Errno 13] Permission denied: '/home/jovyan/.jupyter/migrated'
```

blankenberg committed:
Don't set user in example interactive tool job_conf.xml

blankenberg committed:
Merge branch 'dev' of https://github.com/galaxyproject/galaxy into realtimetools

blankenberg commented:
> OK so it was a configuration problem. I used the documentation in `doc/source/admin/special_topics/` and I added `<tool file="../test/functional/tools/realtimetool_juypter_notebook.xml" />` in `tool_conf.xml`.
> The problem was solved when I changed the tool file path to an absolute path. But I don't understand why the relative path doesn't work.

This does sound like a bug. Would you mind opening an issue explaining the cause?

> And now I have a second issue... This seems to be a docker one.
> 

Yes, this looks to be a user-permission issue between the docker container and the docker runner / environment. If you modify the job destination to set the docker user to _empty_ (`<param id="docker_set_user"></param>`) then it should work. I've added a change to the `job_conf.xml.realtime` example config https://github.com/galaxyproject/galaxy/pull/7494/commits/e04e03d772dbe82eab9abf83c26562df84471760 that should fix this. 

A better/real fix is likely needed in the underlying docker container dependency for this tool.

Please let me know how it goes.

rdallet commented:
> Yes, this looks to be a user-permission issue between the docker container and the docker runner / environment. If you modify the job destination to set the docker user to _empty_ (`<param id="docker_set_user"></param>`) then it should work. I've added a change to the `job_conf.xml.realtime` example config [e04e03d](https://github.com/galaxyproject/galaxy/commit/e04e03d772dbe82eab9abf83c26562df84471760) that should fix this.

It solves the problem but another permission issue appear.
```mkdir: cannot create directory ‘./jupyter’: Permission denied```

jmchilton commented:
Can you add ``requires_domain`` (defaulting to ``False``) or something like that to the tools (``entry_point``) to keep a path-base proxy approach (instead of sub-domain based proxy) a potential default requirement for the future?

bgruening committed:
add a bunch of ITs

bgruening committed:
some small changes

blankenberg committed:
Add API_KEY, GALAXY_URL, and fix HISTORY_ID envars

in interactive_tool_juypter_notebook.xml

bgruening committed:
Merge pull request #72 from blankenberg/patch-1

Add API_KEY, GALAXY_URL, and fix HISTORY_ID envars

blankenberg committed:
Fix websockets in interactive_tool_bam_iobio.xml

bgruening committed:
Merge pull request #73 from blankenberg/patch-1

Fix websockets in interactive_tool_bam_iobio.xml

abretaud committed:
fixes for askomics IT

bgruening committed:
Merge pull request #75 from abretaud/it_askoko

Fixes for askomics IT

blankenberg committed:
Update galaxy.yml.realtime example with outputs_to_working_directory: true

blankenberg committed:
Merge branch 'dev' of https://github.com/galaxyproject/galaxy into realtimetools

bgruening committed:
remove complexity from ipython wrapper, not needed for this version

bgruening committed:
enhance documentation for developers

bgruening committed:
add some more comments

blankenberg committed:
cleanup comments

blankenberg committed:
Merge branch 'dev' of https://github.com/galaxyproject/galaxy into realtimetools

blankenberg committed:
update migration script version

blankenberg committed:
Merge pull request #8 from bgruening/it

migrating a few more interactive environments to interactive tools

jmchilton commented on test/functional/tools/interactive_tool_juypter_notebook.xml:
```diff
@@ -9,14 +9,9 @@
-            #end if
-        </environment_variable>
-        <environment_variable name="REMOTE_HOST">TODO</environment_variable>
+        <environment_variable name="HISTORY_ID" strip="True">${__app__.security.encode_id($jupyter_notebook.history_id)}</environment_variable>
```
We really need a way to inject this and something comparable to ``galaxy_infrastructure_url`` without ``__app__``. Willing to implement higher-level primitives for encoded history ID and inject ``galaxy_infrastructure_url`` with a per-job destination override option (https://github.com/galaxyproject/galaxy/pull/1688/commits/81b5ca564ac7cdd855cb74dad8e98abc83a8efdc). Or at least a warning in every sample that uses ``__app__`` that we will break backward compatibility in the near future

blankenberg replied:
Added warning in 725016230ef56f1e9f25d3018f3643dbfcc539cf and 34ec3bfd3a76a852dcbc9e637af395501e23a5ba

blankenberg committed:
Merge branch 'dev' of https://github.com/galaxyproject/galaxy into realtimetools

blankenberg committed:
Make InteractiveTools enable-able.

blankenberg committed:
minor typo

blankenberg committed:
interactive_tool_juypter_notebook.xml -> interactive_tool_jupyter_notebook.xml

blankenberg committed:
Add requires_domain to entry_point. Default is False. Currently only True works.

blankenberg committed:
Set requires_domain="True" for all examples.

blankenberg committed:
Add warning on backwards compatibility of __app__ in future.

blankenberg committed:
FIXME note on user's API key in tool.

blankenberg committed:
Update galaxy tool XSD.

blankenberg committed:
more realtime to interactivetools name changes

blankenberg committed:
more realtimetool --> interactivetool

blankenberg committed:
more realtimetool --> interactivetool

blankenberg committed:
more realtimetool --> interactivetool

blankenberg commented:
> Can you add `requires_domain` (defaulting to `False`) or something like that to the tools (`entry_point`) to keep a path-base proxy approach (instead of sub-domain based proxy) a potential default requirement for the future?

Added in 41d7155fb6defa3ec83fdaaa93182b9935b6f53d

blankenberg committed:
Fix get_hex for uuid4 in python3

jmchilton commented:
We should use ``false`` not ``False`` in XML probably (https://www.w3schools.com/xml/schema_dtypes_misc.asp), but yeah - looks great, thanks for adding ``requires_domain``!

This and the warnings and them being off by default really assuage a lot of my fears. I think I'm very close to being +1 on this for 19.09. I really am sorry for saying I would merge it sooner when I wasn't ready to and didn't have the time I thought I did to put into it.

Update: Trying to figure out the tests - are they transient problems?

Update 2: Last laugh is yours though because you cause me physical pain when you demo these being used in workflows or tell people to put them on the ToolShed. If I could somehow make my +1 contingent on you not doing those two things - I would - but I don't think that would be feasible. Despite that though - really fantastic work - I think many people are really excited for these.

bgruening commented:
Thanks John for merging.

lizhi16 commented:
The Interactive Tool is an amazing function. Our team is very interested in it. But we don't find any tutorial about how to deploy the InteractiveTools or how to use it. Could you please provide some advice for us to instruct how to use it. Thanks!

bgruening commented:
@lizhi16 we added some initial docs here: https://github.com/blankenberg/galaxy/blob/aea9ed2a219f35ac8b94d11b51ca52ec73cdcb48/doc/source/admin/special_topics/realtimetools.rst

Let us know if this does not help.
